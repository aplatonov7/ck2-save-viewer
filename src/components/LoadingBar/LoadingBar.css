.LoadingBar__container {
  width: 100%;
  height: .5rem;
  overflow: hidden;
  background-color: #c5cae9;
}

.LoadingBar__indicator {
  width: 100%;
  height: 100%;
  background-color: #3f51b5;
  transition: transform 0.2s ease;
}