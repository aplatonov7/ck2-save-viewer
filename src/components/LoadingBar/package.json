{
  "name": "LoadingBar",
  "version": "0.0.0",
  "private": true,
  "main": "./LoadingBar.js"
}