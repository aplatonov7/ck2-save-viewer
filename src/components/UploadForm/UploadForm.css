.UploadForm__container {
  box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24);
  background-color: #fff;
}

.UploadForm__container-loading {
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;

  padding: 2rem;
}

.UploadForm__label {
  display: block;
  padding: 2rem;
  cursor: pointer;
}

.UploadForm__input {
  display: none;
}

.UploadForm__caption {
  margin: 1rem 1rem .5rem;
}

.UploadForm__sub-caption {
  margin: 0 1rem 1rem;
}