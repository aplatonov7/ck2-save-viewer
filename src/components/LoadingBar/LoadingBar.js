import React from 'react'
import PropTypes from 'prop-types'

import './LoadingBar.css'

const LoadingBar = ({ value }) => (
  <div className="LoadingBar__container">
    <div
      className="LoadingBar__indicator"
      style={{ transform: `translateX(${-100 + (value * 100)}%)` }}
    />
  </div>
)

LoadingBar.propTypes = {
  value: PropTypes.number,
}

LoadingBar.defaultProps = {
  value: 0,
}

export default LoadingBar
