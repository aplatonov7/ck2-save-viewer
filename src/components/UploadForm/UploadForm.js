import React, { Component } from 'react'
import PropTypes from 'prop-types'
import cn from 'classnames'

import LoadingBar from '../LoadingBar'
import dataManager from '../../dataManager'

import './UploadForm.css'

class UploadForm extends Component {
  constructor(...props) {
    super(...props)

    this.state = {
      loading: false,
      progress: null,
      filepath: null,
    }
  }

  onProgress = progress => this.setState({ progress })

  readFile = e => {
    const file = e.target.files[0]
    if (!file) return

    const filepath = e.target.value.split('\\').slice(-1)
    this.setState(() => ({ loading: true, filepath }))

    const reader = new FileReader()
    reader.onload = this.handleLoad
    reader.readAsText(file)
  }

  handleLoad = e => {
    const content = e.target.result
    dataManager.loadData(content, this.onProgress).then(() => {
      this.setState(() => ({ loading: false, progress: null, filepath: null }))
      if (this.props.onLoad) this.props.onLoad()
    })
  }

  render() {
    const { loading, progress, filepath } = this.state
    return (
      <div className={cn('UploadForm__container', { 'UploadForm__container-loading': loading })}>
        {this.state.loading ? (
          [
            <div key="1" className="UploadForm__caption">
              Uploading and parsing {filepath}
            </div>,
            <small key="2" className="UploadForm__sub-caption">
              This might take quite a while (-_-) zzz
            </small>,
            <LoadingBar key="3" value={progress} />,
          ]
        ) : (
          <label className="UploadForm__label" htmlFor="saveInput">
            Click to upload your save file
            <input
              className="UploadForm__input"
              id="saveInput"
              type="file"
              onChange={this.readFile}
            />
          </label>
        )}
      </div>
    )
  }
}

UploadForm.propTypes = {
  onLoad: PropTypes.func.isRequired,
}

export default UploadForm
