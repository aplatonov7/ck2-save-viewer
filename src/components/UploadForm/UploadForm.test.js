import React from 'react'
import UploadForm from './UploadForm'

import LoadingBar from '../LoadingBar'

describe('UploadForm component', () => {
  let wrapper

  const props = {
    onLoad: jest.fn(),
  }

  beforeEach(() => {
    wrapper = shallow(<UploadForm {...props} />)
  })

  it('should have proper initial state', () => {
    expect(wrapper.state()).toEqual({ loading: false, filepath: null, progress: null })
  })

  it('should only render input with type=file when state.loading is false', () => {
    wrapper.setState({ loading: false })
    expect(wrapper.find('#saveInput').length).toEqual(1)
    wrapper.setState({ loading: true })
    expect(wrapper.find('#saveInput').length).toEqual(0)
  })

  it('should only render LoadingBar when state.loading is true', () => {
    wrapper.setState({ loading: false })
    expect(wrapper.find(LoadingBar).length).toEqual(0)
    wrapper.setState({ loading: true })
    expect(wrapper.find(LoadingBar).length).toEqual(1)
  })

  it('should match snapshot', () => {
    const tree = renderer.create(<UploadForm {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
