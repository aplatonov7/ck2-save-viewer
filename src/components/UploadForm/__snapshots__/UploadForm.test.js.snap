// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadForm component should match snapshot 1`] = `
<div
  className="UploadForm__container"
>
  <label
    className="UploadForm__label"
    htmlFor="saveInput"
  >
    Click to upload your save file
    <input
      className="UploadForm__input"
      id="saveInput"
      onChange={[Function]}
      type="file"
    />
  </label>
</div>
`;
